test/server/server.test.js

```diff
@@ -0,0 +1,233 @@
+const { graphql } = require("graphql");
+const { agent } = require("supertest");
+const { Schema, root } = require("../../server/gql/schema");
+const { run, stop } = require("../../server/server");
+const { reset, sync } = require("../../server/models");
+const {
+    CREATE_DATA_SOURCE,
+    GET_DATA_SOURCES,
+    GET_DATA_SOURCE,
+    DELETE_DATA_SOURCE,
+    UPDATE_DATA_SOURCE,
+    GET_SCHEMAS,
+    GET_SCHEMA,
+    UPDATE_SCHEMA,
+    GET_RESOLVERS,
+    UPSERT_RESOLVER
+} = require("./queries");
+
+let user = null;
+
+beforeAll(done => {
+    run(app => {
+        user = agent(app);
+        done();
+    });
+});
+
+afterAll(done => {
+    stop(() => done());
+});
+
+afterEach(async () => {
+    await reset();
+    await sync();
+});
+
+describe("Routes", () => {
+    it("should serve the ui", async () => {
+        const res = await user.get("/");
+        expect(res.status).toEqual(200);
+    });
+
+    it("should expose a health endpoint", async () => {
+        const res = await user.get("/healthz");
+        expect(res.status).toEqual(200);
+    });
+
+    // FIXME: receiving 400 instead of 200
+    it.skip("should expose a graphql console", async () => {
+        const res = await user.get("/graphql");
+        expect(res.status).toEqual(200);
+    });
+});
+
+describe("Database", () => {
+    const dataSource = {
+        name: "test",
+        type: "Postgres",
+        config: { timestamps: true },
+        resolvers: []
+    };
+
+    const resolver = {
+        type: "type",
+        field: "field",
+        requestMapping: "",
+        responseMapping: ""
+    };
+
+    describe("DataSources", () => {
+        it("should not have any data sources by default", async () => {
+            const fetch = await graphql(Schema, GET_DATA_SOURCES, root);
+            expect(fetch.data.dataSources).toHaveLength(0);
+        });
+
+        it("should create a new data source", async () => {
+            await graphql(Schema, CREATE_DATA_SOURCE, root, null, dataSource);
+
+            const fetch = await graphql(Schema, GET_DATA_SOURCES, root);
+            expect(fetch.data.dataSources).toHaveLength(1);
+            expect(fetch.data.dataSources[0]).toEqual({ id: 1, ...dataSource });
+        });
+
+        it("should delete an existing data source", async () => {
+            await graphql(Schema, CREATE_DATA_SOURCE, root, null, dataSource);
+
+            let fetch = await graphql(Schema, GET_DATA_SOURCES, root);
```

Comment:
I'm wondering if we could make use of (and that means also test) returned values from the mutations instead of repeating `GET_*` queries in tests?
i.e.
```
const fetch = await graphql(Schema, CREATE_DATA_SOURCE, root, null, dataSource);
expect(fetch.data.createDataSource).toEqual({ id: 1, ...dataSource });
```

Reply:
That is correct and it's actually how I was doing it before. However I think it's more accurate to tests against the DB final state. I might happen that the mutation returned some value that hasn't been added in the DB as we expected. In any case I would check *both*, are those mutation results normally used?

Reply:
I think it generally makes sense to use returned values from performed mutation it in the app. Just my assumption (correct me if I'm wrong), but I see two benefits:
1. Amount or requests made by client app is minimised
2. Content is fetched faster (DB response time could be slow)

Anyway, testing both scenarios seems like a good idea to me.

Reply:
@psturc i'm sure there are situations where using the return value of a mutation instead of an extra request makes sense. But we don't really do that very often and Apollo seems to encourage a different way. For example after adding a new data source, instead of adding the result of the mutation manually and re-rendering, we just use Apollos `refetchQuey` option. Performance wise it's fine: it doesnt fetch anything from the backend because the result of the last query is still in the cache and the mutation also updated the cache with the new data source.

And from a testing perspective we don't really care how many requests are made. Although i agree that it would make sense to at least check that the return values of the mutations are the ones we expect.

Reply:
@pb82 ok thanks for explanation. So since we're not using mutation results anyhow in our app, it's probably not necessary to test them